### DISCUSSION TITLE
What is "Run with args" meant to do?

### DISCUSSION THREAD

**Discussion by LorenzoBettini** [LazyVim/LazyVim#4604] [Q&A]
I've just started using LazyVim for Java development using its Extra and the DAP Extra.

Looking for a way to start a program, e.g., with `<leader> s k` and typing "Run", I got to "Run with args". When asked to specify the args I thought they'd be the arguments to pass to the program, but I guess it's not the case. In particular, passing anything leads to the behavior described here https://github.com/LazyVim/LazyVim/discussions/2414.

In particular, https://github.com/LazyVim/LazyVim/discussions/2414#discussioncomment-10882814 seems to also imply that "args" were used with the above described semantics, which I think it's wrong.

I seem to understand that `continue` is the right way to start a program (I also created https://github.com/LazyVim/LazyVim/pull/4603 concerning `continue`)

**Comment by rogeriofrsouza** [comment]
These are indeed args to the program.

There are some docs provided by [nvim-dap](https://github.com/mfussenegger/nvim-dap) for continue function in `<leader> s h dap-api`

>         If no debug session is active, `continue()` will start a new debug session
>         [...]
>         `continue()` is the main entry-point for users to start debugging an application.

Do you have a nvim-dap debugger configuration for java?
It would be something like this:
```lua
{
    "mfussenegger/nvim-dap",
    opts = function()
      -- https://github.com/mfussenegger/nvim-dap/wiki/Java
      local dap = require("dap")
      dap.configurations.java = {
        {
          type = "java",
          request = "attach",
          name = "Debug (Attach) - Remote",
          hostName = "127.0.0.1",
          port = 5005,
        },
      }
    end,
    ```

**Reply by LorenzoBettini** [reply]
@rogeriofrsouza thank you for the answer, but I still don't understand ;)

First, it is clear that `continue`, without arguments, has to be used to start a new debug session (I only find misleading the term `continue` also to imply "start", but that's another issue, and I created a PR to hopefully help to understand its semantics better, https://github.com/LazyVim/LazyVim/pull/4603)

I still don't understand "run with args". Looking at what the keybinding does

https://github.com/LazyVim/LazyVim/blob/v12.43.0/lua/lazyvim/plugins/extras/dap/core.lua#L34

```lua
{ "<leader>da", function() require("dap").continue({ before = get_args }) end, desc = "Run with Args" },
```

it passes to `continue` a table with key `before` and value the result of `get_args` (please excuse me if I got it wrong; I don't know Lua much)

https://github.com/LazyVim/LazyVim/blob/v12.43.0/lua/lazyvim/plugins/extras/dap/core.lua#L2

```lua
local function get_args(config)
  local args = type(config.args) == "function" and (config.args() or {}) or config.args or {}
  config = vim.deepcopy(config)
  ---@cast args string[]
  config.args = function()
    local new_args = vim.fn.input("Run with args: ", table.concat(args, " ")) --[[@as string]]
    return vim.split(vim.fn.expand(new_args) --[[@as string]], " ")
  end
  return config
end
```

so the function doesn't get anything because it's called without arguments (I seem to understand), and it returns the string inserted in the popup dialog (or a table with that string?).

In any case, the `continue(...)` that takes an argument

https://github.com/mfussenegger/nvim-dap/blob/master/doc/dap.txt#L681

```
continue({opts})                                                *dap.continue()*
        `continue()` resumes the execution of an application [count] times if a
        debug session is active and a thread was stopped. Threads are usually
        stopped when a breakpoint is hit or an exception occurred.

        If no debug session is active, `continue()` will start a new debug session by:

        - Looking up the configurations (|dap-configuration|) for the current filetype.
        - If there is more than one configuration it will prompt the user to
          select one of them.
        - It calls |dap.run()| on the selected configuration.

        `continue()` is the main entry-point for users to start debugging an
        application.

        Parameters:
            {opts}    Optional table with:
                      - `new: boolean` force starting an additional debug session
```

Says that it expects a table with key `new: boolean`, so how can it be used to pass arguments to the program to debug?

Concerning configurations, the Java extra already creates configurations on the fly, looking at Java files with `main` methods: https://github.com/LazyVim/LazyVim/blob/v12.43.0/lua/lazyvim/plugins/extras/lang/java.lua#L231 so I didn't provide any custom configurations.

Looking at other API of dap, e.g., https://github.com/mfussenegger/nvim-dap/blob/master/doc/dap.txt#L701, I seem to understand that `run({config}, {opts})` can be used, but again the arguments you would pass to such a function don't seem to imply that you pass directly arguments to the program you want to debug.

Again, apologizes if I got anything wrong.

**Reply by dpetka2001** [reply]
You should actually look at [`before`'s function signature here](https://github.com/mfussenegger/nvim-dap/blob/7ff6936010b7222fea2caea0f67ed77f1b7c60dd/lua/dap.lua#L559). It takes a `dap.Configuration` argument and `dap.Configuration` can also include `args` that will get passed as arguments to the corresponding configuration for the adapter.

See [here](https://github.com/mfussenegger/nvim-dap/blob/7ff6936010b7222fea2caea0f67ed77f1b7c60dd/doc/dap.txt#L61-L65) an example for adapter configuration (`debugpy` in this case). It accepts `args` inside the Dap Configuration for the adapter. So, `Run with args` just means that it will run the adapter configuration with the `args` that the user will provide. 

I don't use DAP myself, so can't be of further help with regards to its usage.

**Reply by LorenzoBettini** [reply]
@dpetka2001 but `continue` that accepts an argument just expects a table with `new` key, see the documentation above; `continue` with an argument doesn't expect a configuration... how can this work?

Moreover, I still don't understand how the string inserted in the input field (see the `get_args` above) is turned into a value of the key `args`.

**Reply by dpetka2001** [reply]
> run({config}, {opts})                                                *dap.run()*
        Looks up a debug adapter entry for the given configuration and runs it.
        This is implicitly called by |dap.continue()| if no debug session is
        active.

**Reply by LorenzoBettini** [reply]
Moreover, look at the definition of `continue`, https://github.com/mfussenegger/nvim-dap/blob/master/lua/dap.lua#L1019

```lua
---@param opts? {new?: boolean}
function M.continue(opts)
  if not session then
    session = first_stopped_session()
  end

  opts = opts or {}
  if not session or opts.new then
    select_config_and_run(opts)
  elseif session.stopped_thread_id then
...
```

It never uses the `before` field; as stated in the documentation, it only uses the optional `new` field.

It is `run(config, opts)`, https://github.com/mfussenegger/nvim-dap/blob/master/lua/dap.lua#L565, which understands and uses `before` in `opts`

```lua
--- Start a debug session
---@param config dap.Configuration
---@param opts dap.run.opts?
function M.run(config, opts)
  assert(
    type(config) == 'table',
    'dap.run() must be called with a valid configuration, got ' .. vim.inspect(config))

  opts = opts or {}
  if session and (opts.new == false or (opts.new == nil and session.config.name == config.name)) then
    M.restart(config, opts)
    return
  end
  opts.filetype = opts.filetype or vim.bo.filetype
  opts.new = nil
  last_run = {
    config = config,
    opts = opts,
  }
  if opts.before then
    config = opts.before(config)
  end
...
```

**Reply by dpetka2001** [reply]
`continue` passes its `opts` to `select_config_and_run(opts)` as you can see 
```lua
if not session or opts.new then
    select_config_and_run(opts)
```
So, it seems you can pass `opts` that `M.run(config, opts)` will understand.

**Reply by LorenzoBettini** [reply]
I did a few experiments.

I tried with ".vscode/launch.json", which is supported, of this shape:

```json
{
	"$schema": "https://raw.githubusercontent.com/mfussenegger/dapconfig-schema/master/dapconfig-schema.json",
	"version": "0.2.0",
	"configurations": [
		{
			"type": "java",
			"request": "launch",
			"name": "Launch Java"
		}
	]
}
```

and "Run with args" and nothing happens: the debugging is blocked like reported here https://github.com/LazyVim/LazyVim/discussions/2414#discussioncomment-10882814.

I tried with this, with `args`:

```json
{
	"$schema": "https://raw.githubusercontent.com/mfussenegger/dapconfig-schema/master/dapconfig-schema.json",
	"version": "0.2.0",
	"configurations": [
		{
			"type": "java",
			"request": "launch",
			"name": "Launch Java",
			"args": "first second"
		}
	]
}
```

and now "Run with args" (after selecting this configuration) does not even get to ask for the arguments:

```
notify.error DAP ...ava/lazy/LazyVim/lua/lazyvim/plugins/extras/dap/core.lua:7: bad argument #1 to 'concat' (table expected, got string)
stack traceback:
	[C]: in function 'concat'
	...ava/lazy/LazyVim/lua/lazyvim/plugins/extras/dap/core.lua:7: in function 'option'
	...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:307: in function 'eval_option'
	...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:363: in function <...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:362>
	vim/shared.lua: in function 'on_config'
	...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:551: in function 'prepare_config'
	...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:585: in function <...tini/.local/share/lazyvim-java/lazy/nvim-dap/lua/dap.lua:584>
	[C]: in function 'xpcall'
	...local/share/lazyvim-java/lazy/nvim-dap/lua/dap/async.lua:12: in function <...local/share/lazyvim-java/lazy/nvim-dap/lua/dap/async.lua:11>
```

In fact, the `get_args` assumes that the possibly existing `args` is a table, while here it's a string, and https://raw.githubusercontent.com/mfussenegger/dapconfig-schema/master/dapconfig-schema.json shows that for type "java", "args" must be a string.

I then tried to manually run `run( config opts )` (calling directly `continue(opts)` would not work because it doesn't pass through the `before` thing); I tried with a Java main file printing the command line arguments.

Here are the results:

```
:lua require("dap").run({type = "java", request = "launch", name = "Custom", args = "foo bar"})
```

works correctly: the arguments are passed.

However, passing arguments as a table:

```
:lua require("dap").run({type = "java", request = "launch", name = "Custom", args = {"foo, bar"}})
```

or even an empty table

```
:lua require("dap").run({type = "java", request = "launch", name = "Custom", args = {}})
```

blocks the debugging once again:

```
Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or `java` configuration. Check the logs for errors (:help dap.set_log_level)
```

By the way, increasing the debugging level doesn't show anything on the `dap.log`.

I can only come to the conclusion that the Java adapter cannot handle `args` as a table, but only as a string.

**Reply by dpetka2001** [reply]
Like I said I don't use `nvim-dap` and also don't do Java. The Java Extra was created by someone who did Java.

I see in the Java Extra that it overwrites the `test.core` Extra mappings when `jdtls` attaches [here](https://github.com/LazyVim/LazyVim/blob/cb40a09538dc0c417a7ffbbacdbdec90be4a792c/lua/lazyvim/plugins/extras/lang/java.lua#L236-L261). Maybe you could do something similar for the `Run with args`, so that it takes into account that `args` is a string. Because like you said function `get_args` in `dap.core` Extra does return `args` as a list of strings instead.

Maybe if you manage to solve it locally you could create a PR to improve the current Java Extra.

**Reply by LorenzoBettini** [reply]
PR created: https://github.com/LazyVim/LazyVim/pull/4673

**Comment by LorenzoBettini** [comment]
It is now working correctly also with Java!
I'm closing this.
